In [ ]:
# Google Cloud Authentication and Setup

!pip install --quiet google-cloud-storage google-cloud-bigquery

from google.colab import auth
auth.authenticate_user()
print("✅ Authenticated to Google Cloud")

✅ Authenticated to Google Cloud


In [ ]:
# Mount Google Drive

from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# Initialize Google Cloud Storage and Create/Get Bucket

from google.cloud import storage

PROJECT_ID = "mgmt467-project"
BUCKET_NAME = "mgmt467-gym-raw"

client = storage.Client(project=PROJECT_ID)

try:
    bucket = client.get_bucket(BUCKET_NAME)
    print("Bucket already exists:", BUCKET_NAME)
except:
    bucket = client.create_bucket(BUCKET_NAME, location="US")
    print("Created bucket:", BUCKET_NAME)

Bucket already exists: mgmt467-gym-raw


In [ ]:
# Export BigQuery Table to Cloud Storage

from google.cloud import bigquery

bq = bigquery.Client(project=PROJECT_ID)

DATASET = "GymDB"
SOURCE_TABLE = "GymData"

DEST_URI = f"gs://{BUCKET_NAME}/batch/GymData_export.csv"

extract_job = bq.extract_table(
    f"{PROJECT_ID}.{DATASET}.{SOURCE_TABLE}",
    DEST_URI,
    location="US"
)

extract_job.result()
print("✅ Exported GymData to:", DEST_URI)

✅ Exported GymData to: gs://mgmt467-gym-raw/batch/GymData_export.csv


In [ ]:
# Load Raw Data from Cloud Storage to BigQuery

RAW_TABLE = "GymData_Raw"
uri = f"gs://{BUCKET_NAME}/batch/GymData_export.csv"

job_config = bigquery.LoadJobConfig(
    source_format=bigquery.SourceFormat.CSV,
    skip_leading_rows=0,  # BigQuery export has no header row
    autodetect=True
)

load_job = bq.load_table_from_uri(
    uri,
    f"{PROJECT_ID}.{DATASET}.{RAW_TABLE}",
    job_config=job_config
)

load_job.result()
print("✅ Loaded raw table:", f"{PROJECT_ID}.{DATASET}.{RAW_TABLE}")

✅ Loaded raw table: mgmt467-project.GymDB.GymData_Raw


In [ ]:
# Create or Replace Raw Data Table in BigQuery

query = f"""
CREATE OR REPLACE TABLE `mgmt467-project.GymDB.GymData_Raw` AS
SELECT *
FROM `mgmt467-project.GymDB.GymData`;
"""

bq.query(query).result()

In [ ]:
CURATED_TABLE = "GymData_Curated"

query = f"""
CREATE OR REPLACE TABLE `mgmt467-project.GymDB.GymData_Curated`
PARTITION BY Last_Visit_Date AS
SELECT
  *,
  DATE_DIFF(Last_Visit_Date, Join_Date, DAY) AS Tenure_Days,
  IF(Churn = TRUE, 1, 0) AS Churn_Label
FROM `mgmt467-project.GymDB.GymData_Raw`;
"""

bq.query(query).result()
print("✅ Created curated table:", f"{PROJECT_ID}.{DATASET}.{CURATED_TABLE}")

✅ Created curated table: mgmt467-project.GymDB.GymData_Curated


In [ ]:
dq_query = """
SELECT
  COUNT(*) AS total_rows,
  COUNTIF(Member_ID IS NULL) AS missing_member_id,
  COUNTIF(Last_Visit_Date IS NULL) AS missing_last_visit_date,
  COUNTIF(Join_Date IS NULL) AS missing_join_date,
  COUNTIF(Age IS NULL) AS missing_age,
  MIN(Age) AS min_age,
  MAX(Age) AS max_age
FROM `mgmt467-project.GymDB.GymData_Curated`;
"""

dq_result = bq.query(dq_query).result().to_dataframe()

print("===== DATA QUALITY CHECK RESULTS =====")
print(dq_result)

===== DATA QUALITY CHECK RESULTS =====
   total_rows  missing_member_id  missing_last_visit_date  missing_join_date  \
0         150                  0                        0                  9   

   missing_age  min_age  max_age  
0           13     18.0     54.0  


In [ ]:
transform_query = """
SELECT
  Member_ID,
  Join_Date,
  Last_Visit_Date,
  DATE_DIFF(Last_Visit_Date, Join_Date, DAY) AS Tenure_Days,
  Churn,
  IF(Churn = TRUE, 1, 0) AS Churn_Label
FROM `mgmt467-project.GymDB.GymData_Curated`
LIMIT 10;
"""

transform_result = bq.query(transform_query).result().to_dataframe()

print("===== TRANSFORMATION LOGIC PREVIEW =====")
print(transform_result)

===== TRANSFORMATION LOGIC PREVIEW =====
   Member_ID   Join_Date Last_Visit_Date  Tenure_Days  Churn  Churn_Label
0         41  2023-08-28      2024-01-02          127  False            0
1        128  2023-05-27      2023-08-16           81   True            1
2        120  2023-07-28      2023-08-16           19  False            0
3        124         NaT      2022-09-23         <NA>  False            0
4         99  2022-05-10      2022-12-07          211  False            0
5         14  2024-08-23      2025-06-29          310  False            0
6         58         NaT      2025-02-22         <NA>   True            1
7         31  2024-07-17      2025-02-22          220   True            1
8         44  2024-03-16      2025-03-10          359  False            0
9         92  2024-01-16      2024-04-13           88  False            0


In [ ]:
from google.colab import auth
auth.authenticate_user()

from google.cloud import bigquery

PROJECT_ID = "mgmt467-project"

# IMPORTANT: This MUST be named bq — and DO NOT overwrite it later
bq = bigquery.Client(project=PROJECT_ID)

print("BigQuery client reset. Connected to:", PROJECT_ID)

BigQuery client reset. Connected to: mgmt467-project


In [ ]:
bq.query("DROP TABLE IF EXISTS `mgmt467-project.GymDB.Weather_Raw_Streaming`;").result()
bq.query("DROP TABLE IF EXISTS `mgmt467-project.GymDB.Weather_Monthly_Avg`;").result()

# Create new raw streaming table
bq.query("""
CREATE TABLE `mgmt467-project.GymDB.Weather_Raw_Streaming`
(
  event_timestamp TIMESTAMP,
  temperature_c FLOAT64,
  wind_speed_kph FLOAT64,
  relative_humidity FLOAT64,
  is_day BOOL
)
PARTITION BY DATE(event_timestamp)
""").result()

# Create monthly aggregated table
bq.query("""
CREATE TABLE `mgmt467-project.GymDB.Weather_Monthly_Avg`
(
  year INT64,
  month INT64,
  avg_temperature_c FLOAT64,
  avg_wind_speed_kph FLOAT64,
  avg_relative_humidity FLOAT64
)
""").result()

print("Tables recreated successfully.")

Tables recreated successfully.


In [ ]:
from google.colab import auth
auth.authenticate_user()

from google.cloud import bigquery

PROJECT_ID = "mgmt467-project"   # <-- USE NAME, NOT NUMBER
DATASET_ID = "GymDB"

bq = bigquery.Client(project=PROJECT_ID)

print("Authenticated project:", bq.project)

Authenticated project: mgmt467-project


In [ ]:
from google.cloud import bigquery

PROJECT_ID = "mgmt467-project"
DATASET_ID = "GymDB"
TABLE_ID = "Weather_Raw_Streaming"

# -----------------------
# 1) Ensure dataset exists
# -----------------------
dataset_ref = bigquery.Dataset(f"{PROJECT_ID}.{DATASET_ID}")
dataset = bq.create_dataset(dataset_ref, exists_ok=True)
print("Dataset verified:", DATASET_ID)

# -----------------------
# 2) Delete table if exists
# -----------------------
try:
    bq.delete_table(f"{PROJECT_ID}.{DATASET_ID}.{TABLE_ID}")
    print("Old table deleted.")
except:
    print("No existing table, skipping delete.")

# -----------------------
# 3) Recreate table with schema
# -----------------------
schema = [
    bigquery.SchemaField("event_timestamp", "TIMESTAMP"),
    bigquery.SchemaField("temperature_c", "FLOAT"),
    bigquery.SchemaField("wind_speed_kph", "FLOAT"),
    bigquery.SchemaField("relative_humidity", "FLOAT"),
    bigquery.SchemaField("is_day", "BOOL"),
]

table_ref = bigquery.Table(
    f"{PROJECT_ID}.{DATASET_ID}.{TABLE_ID}",
    schema=schema
)

table = bq.create_table(table_ref)
print("Table created:", table.full_table_id)

Dataset verified: GymDB
Old table deleted.
Table created: mgmt467-project:GymDB.Weather_Raw_Streaming


In [ ]:
import requests
import pandas as pd
from datetime import datetime, timedelta

LAT = 40.4237
LON = -86.9212

end_date = datetime.utcnow().date()
start_date = end_date - timedelta(days=365)

url = "https://archive-api.open-meteo.com/v1/archive"

params = {
    "latitude": LAT,
    "longitude": LON,
    "start_date": start_date.isoformat(),
    "end_date": end_date.isoformat(),
    "hourly": "temperature_2m,relativehumidity_2m,windspeed_10m",
    "timezone": "UTC"
}

data = requests.get(url, params=params).json()

df_hist = pd.DataFrame({
    "event_timestamp": data["hourly"]["time"],
    "temperature_c": data["hourly"]["temperature_2m"],
    "wind_speed_kph": data["hourly"]["windspeed_10m"],
    "relative_humidity": data["hourly"]["relativehumidity_2m"],
})

df_hist["is_day"] = pd.to_datetime(df_hist["event_timestamp"]).dt.hour.between(6, 18)

df_hist.head(), df_hist.shape


/tmp/ipython-input-4181387914.py:8: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  end_date = datetime.utcnow().date()


(    event_timestamp  temperature_c  wind_speed_kph  relative_humidity  is_day
 0  2024-12-09T00:00            8.5            17.1                 83   False
 1  2024-12-09T01:00            9.2            16.5                 75   False
 2  2024-12-09T02:00            9.1            19.6                 79   False
 3  2024-12-09T03:00            9.3            18.8                 85   False
 4  2024-12-09T04:00            9.5            20.2                 89   False,
 (8784, 5))

In [ ]:
from google.colab import auth
auth.authenticate_user()

from google.cloud import bigquery

PROJECT_ID = "mgmt467-project"
DATASET = "GymDB"
RAW_TABLE = "Weather_Raw_Streaming"

bq = bigquery.Client(project=PROJECT_ID)

# Drop old table
bq.query(f"DROP TABLE IF EXISTS `{PROJECT_ID}.{DATASET}.{RAW_TABLE}`;").result()

# Recreate raw table schema
bq.query(f"""
CREATE TABLE `{PROJECT_ID}.{DATASET}.{RAW_TABLE}`
(
  event_timestamp TIMESTAMP,
  temperature_c FLOAT64,
  wind_speed_kph FLOAT64,
  relative_humidity FLOAT64,
  is_day BOOL
)
PARTITION BY DATE(event_timestamp)
""").result()

print("Weather_Raw_Streaming table recreated.")



Weather_Raw_Streaming table recreated.


In [ ]:
from google.cloud import bigquery

PROJECT_ID = "mgmt467-project"
DATASET_ID = "GymDB"
TABLE_NAME = "Weather_Raw_Streaming"

# Build SAFE reference (this is the fix)
dataset_ref = bigquery.DatasetReference(PROJECT_ID, DATASET_ID)
table_ref = bigquery.TableReference(dataset_ref, TABLE_NAME)

rows = df_hist.to_dict(orient="records")

batch_size = 500
errors_total = []

for i in range(0, len(rows), batch_size):
    batch = rows[i:i+batch_size]

    errors = bq.insert_rows_json(table_ref, batch)  # << fixed line

    if errors:
        print("Batch insertion errors example:", errors[:3])
    errors_total.extend(errors)

print("Done! Total insertion errors:", len(errors_total))


Done! Total insertion errors: 0


In [ ]:
MONTHLY_TABLE = "Weather_Monthly_Avg"

# Drop old monthly table
bq.query(f"DROP TABLE IF EXISTS `{PROJECT_ID}.{DATASET}.{MONTHLY_TABLE}`;").result()

# Create aggregated table
bq.query(f"""
CREATE TABLE `{PROJECT_ID}.{DATASET}.{MONTHLY_TABLE}` AS
SELECT
  EXTRACT(YEAR FROM event_timestamp) AS year,
  EXTRACT(MONTH FROM event_timestamp) AS month,
  AVG(temperature_c) AS avg_temperature_c,
  AVG(wind_speed_kph) AS avg_wind_speed_kph,
  AVG(relative_humidity) AS avg_relative_humidity
FROM `{PROJECT_ID}.{DATASET}.{RAW_TABLE}`
GROUP BY year, month
ORDER BY year, month
""").result()

print("Monthly averages computed.")


Monthly averages computed.


In [ ]:
df_monthly = bq.query(f"""
SELECT *
FROM `{PROJECT_ID}.{DATASET}.Weather_Monthly_Avg`
ORDER BY year, month
""").to_dataframe()

df_monthly

year  month  avg_temperature_c  avg_wind_speed_kph  avg_relative_humidity
0   2024     12           2.690399           13.582971              80.458333
1   2025      1          -5.408602           14.267742              71.705645
2   2025      2          -1.235863           13.448958              69.824405
3   2025      3           8.181989           16.437769              61.850806
4   2025      4          11.602778           14.082778              66.873611
5   2025      5          16.089113           12.454301              68.994624
6   2025      6          23.355417           11.228889              71.362500
7   2025      7          25.546774            8.716263              75.069892
8   2025      8          22.348790            9.046237              71.958333
9   2025      9          20.681806            7.891528              61.745833
10  2025     10          14.623118           12.119489              62.288978
11  2025     11           6.221389           13.405000              68.350000
12  2025     12          -5.148611           10.998148              86.875000

In [ ]:
from google.cloud import bigquery
from datetime import datetime, timezone
import requests, time

PROJECT_ID = "mgmt467-project"
DATASET_ID = "GymDB"
TABLE_NAME = "Weather_Raw_Streaming"

bq = bigquery.Client(project=PROJECT_ID)

dataset_ref = bigquery.DatasetReference(PROJECT_ID, DATASET_ID)
table_ref = bigquery.TableReference(dataset_ref, TABLE_NAME)

In [ ]:
def fetch_weather_live():
    url = "https://api.open-meteo.com/v1/forecast"
    params = {
        "latitude": 40.4237,
        "longitude": -86.9212,
        "current_weather": True
    }
    d = requests.get(url, params=params).json()
    cw = d["current_weather"]

    return {
        "event_timestamp": datetime.now(timezone.utc).isoformat(),
        "temperature_c": cw["temperature"],
        "wind_speed_kph": cw["windspeed"],
        "relative_humidity": None,  # live endpoint does not include humidity in all cases
        "is_day": bool(cw["is_day"])
    }

In [ ]:
print("Starting live streaming... Press STOP to end.")

for i in range(20):  # change to while True for nonstop streaming
    row = fetch_weather_live()
    errors = bq.insert_rows_json(table_ref, [row])

    print(f"[{i}] Inserted:", row, "| Errors:", errors)

    time.sleep(60)  # 1 reading per minute

Starting live streaming... Press STOP to end.
[0] Inserted: {'event_timestamp': '2025-12-09T20:39:15.419571+00:00', 'temperature_c': 4.2, 'wind_speed_kph': 24.7, 'relative_humidity': None, 'is_day': True} | Errors: []


KeyboardInterrupt: 

# Task
Create and train a BigQuery ML logistic regression model named `mgmt467-project.GymDB.churn_prediction_model` using a joined dataset of `mgmt467-project.GymDB.GymData_Curated` and aggregated daily weather data from `mgmt467-project.GymDB.Weather_Raw_Streaming`. Then, evaluate its performance using `ML.EVALUATE` and analyze feature importances with `ML.FEATURE_IMPORTANCES`, specifically focusing on the impact of weather-related features on churn prediction.

## Setup BigQuery Client and Constants

### Subtask:
Initialize the BigQuery client and define the project ID, dataset ID, and the names of the `GymData_Curated` and `Weather_Raw_Streaming` tables. Also, define a name for the new BQML model we will create.


**Reasoning**:
First, I'll authenticate the user to Google Cloud, import the BigQuery client, and then initialize the client with the project ID. Afterwards, I will define all the necessary constants for the project, including the project ID, dataset ID, curated gym data table, raw streaming weather table, and the new BQML model name. This setup is crucial for subsequent BigQuery operations.



In [ ]:
from google.colab import auth
from google.cloud import bigquery

auth.authenticate_user()
print("✅ Authenticated to Google Cloud")

PROJECT_ID = "mgmt467-project"
DATASET_ID = "GymDB"
GYM_DATA_CURATED_TABLE = "GymData_Curated"
WEATHER_RAW_STREAMING_TABLE = "Weather_Raw_Streaming"
MODEL_NAME = "churn_prediction_model"

bq = bigquery.Client(project=PROJECT_ID)

print(f"✅ BigQuery client initialized for project: {PROJECT_ID}")
print(f"Dataset ID: {DATASET_ID}")
print(f"Gym Data Curated Table: {GYM_DATA_CURATED_TABLE}")
print(f"Weather Raw Streaming Table: {WEATHER_RAW_STREAMING_TABLE}")
print(f"Model Name: {MODEL_NAME}")

✅ Authenticated to Google Cloud
✅ BigQuery client initialized for project: mgmt467-project
Dataset ID: GymDB
Gym Data Curated Table: GymData_Curated
Weather Raw Streaming Table: Weather_Raw_Streaming
Model Name: churn_prediction_model


## Construct BQML Model Training Query

### Subtask:
Formulate a BigQuery SQL query to create and train a BQML logistic regression model. This query will involve: 1. Aggregating the `Weather_Raw_Streaming` table to daily averages (temperature, wind speed, humidity, day/night ratio) to represent streaming features. 2. Joining the `GymData_Curated` table (batch features) with the daily aggregated weather data on `Last_Visit_Date`. 3. Selecting relevant features for the model, with `Churn_Label` as the target variable. BQML will handle one-hot encoding for categorical features and default imputation for missing numerical values.


**Reasoning**:
I need to construct a SQL query string that creates and trains a BigQuery ML logistic regression model. This involves aggregating weather data, joining it with the curated gym data, and selecting relevant features as specified in the subtask.



**Reasoning**:
The `training_query` has been defined; the next step is to execute this query using the BigQuery client to create and train the BigQuery ML model.



In [ ]:
training_query = f"""
CREATE OR REPLACE MODEL `{PROJECT_ID}.{DATASET_ID}.{MODEL_NAME}`
OPTIONS(
    model_type='LOGISTIC_REG',
    input_label_cols=['Churn_Label']
) AS
SELECT
  g.Churn_Label,
  g.Tenure_Days,
  g.Age,
  g.Gender,
  g.Membership_Type,
  w.avg_daily_temperature_c,
  w.avg_daily_wind_speed_kph,
  w.avg_daily_relative_humidity,
  w.day_proportion
FROM
  `{PROJECT_ID}.{DATASET_ID}.{GYM_DATA_CURATED_TABLE}` AS g
LEFT JOIN
  (
    SELECT
      DATE(event_timestamp) AS weather_date,
      AVG(temperature_c) AS avg_daily_temperature_c,
      AVG(wind_speed_kph) AS avg_daily_wind_speed_kph,
      AVG(relative_humidity) AS avg_daily_relative_humidity,
      SUM(CASE WHEN is_day THEN 1 ELSE 0 END) / COUNT(*) AS day_proportion
    FROM
      `{PROJECT_ID}.{DATASET_ID}.{WEATHER_RAW_STREAMING_TABLE}`
    GROUP BY
      weather_date
  ) AS w
ON
  g.Last_Visit_Date = w.weather_date
WHERE
  g.Churn_Label IS NOT NULL -- Only train on records where churn status is known
"""

print("SQL training query defined.")

SQL training query defined.


**Reasoning**:
The `training_query` has been redefined; the next step is to execute this corrected query using the BigQuery client to create and train the BigQuery ML model.



In [ ]:
print("Starting model training...")
bq.query(training_query).result()
print(f"✅ Model '{MODEL_NAME}' training initiated.")

Starting model training...
✅ Model 'churn_prediction_model' training initiated.


## Evaluate BQML Model Performance

### Subtask:
Run a ML.EVALUATE query on the trained BQML model to assess its performance. We will check metrics such as accuracy, precision, recall, and F1-score to understand how well the model predicts customer churn.


**Reasoning**:
I need to construct a SQL query to evaluate the trained BQML model using ML.EVALUATE, execute it, convert the result to a Pandas DataFrame, and print it to display the performance metrics.



In [ ]:
evaluation_query = f"""
SELECT
  *
FROM
  ML.EVALUATE(MODEL `{PROJECT_ID}.{DATASET_ID}.{MODEL_NAME}`)
"""

print("Running model evaluation query...")
eval_result = bq.query(evaluation_query).result().to_dataframe()

print("\n===== MODEL EVALUATION RESULTS =====")
print(eval_result)

Running model evaluation query...

===== MODEL EVALUATION RESULTS =====
   precision    recall  accuracy  f1_score  log_loss   roc_auc
0        1.0  0.025641  0.746667      0.05  0.555046  0.594548


## Analyze Feature Importances

### Subtask:
Analyze feature importances using `ML.FEATURE_IMPORTANCES` to understand which features, particularly weather-related ones, have the most significant impact on the churn prediction model.

#### Instructions:
1. Construct a BigQuery SQL query to retrieve feature importances for the trained model named `churn_prediction_model`.
2. Use the `ML.FEATURE_IMPORTANCES` function, specifying the fully qualified model name.
3. Execute this query using the `bq.query().result()` method.
4. Convert the query result into a Pandas DataFrame using `.to_dataframe()`.
5. Print the resulting DataFrame to display the feature importances, and also display the top N features for a more focused analysis.

**Reasoning**:
The error `Unrecognized name: processed_input_weights` indicates that the column name used for ordering is incorrect in the `ML.WEIGHTS` output. I need to use the `weight` column instead, which represents the feature coefficients for logistic regression models.



In [ ]:
feature_importance_query = f"""
SELECT
  * EXCEPT(category_weights)
FROM
  ML.WEIGHTS(MODEL `{PROJECT_ID}.{DATASET_ID}.{MODEL_NAME}`)
ORDER BY
  ABS(weight) DESC
"""

print("Running feature importance query...")
feature_importance_result = bq.query(feature_importance_query).result().to_dataframe()

print("\n===== FEATURE IMPORTANCE RESULTS =====")
print(feature_importance_result)


Running feature importance query...

===== FEATURE IMPORTANCE RESULTS =====
               processed_input    weight
0                __INTERCEPT__ -0.317823
1      avg_daily_temperature_c -0.059069
2                          Age  0.012715
3     avg_daily_wind_speed_kph -0.011281
4  avg_daily_relative_humidity -0.000162
5                  Tenure_Days -0.000023
6               day_proportion  0.000000
7                       Gender       NaN
8              Membership_Type       NaN


## Final Task

### Subtask:
Summarize the BQML model creation process, its performance metrics, and highlight the key features influencing churn prediction, specifically discussing the contribution of weather-related data.


## Summary:

### Q&A
The task asked for a summary of the BQML model creation process, its performance metrics, and key features influencing churn prediction, specifically discussing the contribution of weather-related data.

### Data Analysis Key Findings
*   A BigQuery ML logistic regression model, `churn_prediction_model`, was successfully created and trained using a joined dataset. This dataset combined `GymData_Curated` (batch features like Tenure, Age, Gender, Membership Type) with daily aggregated weather data from `Weather_Raw_Streaming` (average daily temperature, wind speed, relative humidity, and day proportion).
*   During model creation, an initial error related to a non-existent `Location` column in `GymData_Curated` was identified and corrected by removing the column from the training query.
*   The model's performance metrics were evaluated:
    *   **Accuracy:** 0.746667
    *   **Precision:** 1.0
    *   **Recall:** 0.025641
    *   **F1-Score:** 0.05
    *   **Log Loss:** 0.555046
    *   **ROC AUC:** 0.594548
    These metrics indicate that while the model is very precise when it predicts churn, it misses a significant portion of actual churners (low recall), leading to a poor overall F1-score for the positive class.
*   Feature importances (weights) for the logistic regression model were successfully extracted using `ML.WEIGHTS` after correcting an attempt to use `ML.FEATURE_IMPORTANCES` (which is not applicable to logistic regression).
*   The top features by absolute weight were:
    *   `__INTERCEPT__`
    *   `avg_daily_temperature_c` (-0.059069)
    *   `Age` (0.012715)
    *   `avg_daily_wind_speed_kph` (-0.011281)
*   **Contribution of weather-related data:** Weather-related features (`avg_daily_temperature_c`, `avg_daily_wind_speed_kph`, `avg_daily_relative_humidity`, `day_proportion`) are present among the influential features. Specifically, `avg_daily_temperature_c` is the second most impactful feature after the intercept, with a negative weight. This suggests that higher average daily temperatures are associated with a decreased likelihood of churn. `avg_daily_wind_speed_kph` also shows a notable (negative) impact.

### Insights or Next Steps
*   The model's current performance, particularly its low recall and F1-score for churn prediction, indicates that it is not effectively identifying customers who are likely to churn. Further investigation into the class imbalance, feature engineering, or exploring different model types (e.g., ensemble methods) is warranted to improve churn detection.
*   The significant impact of `avg_daily_temperature_c` on churn prediction suggests that external environmental factors play a role in customer retention. Further analysis could explore specific temperature thresholds or seasonal patterns, and this insight could inform targeted marketing or engagement strategies during periods of expected higher churn due to weather conditions.
